File: Android/dokit-gps-mock/src/main/java/com/didichuxing/doraemonkit/gps_mock/gpsmock/TelephonyHooker.java
package com.didichuxing.doraemonkit.gps_mock.gpsmock;

import android.content.Context;
import android.os.IBinder;
import android.telephony.CellInfo;
import android.telephony.gsm.GsmCellLocation;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by wanglikun on 2019/4/2
 */
public class TelephonyHooker extends BaseServiceHooker {
    @Override
    public String serviceName() {
        return Context.TELEPHONY_SERVICE;
    }

    @Override
    public String stubName() {
        return "com.android.internal.telephony.ITelephony$Stub";
    }

    @Override
    public Map<String, MethodHandler> registerMethodHandlers() {
        Map<String, MethodHandler> methodHandlers = new HashMap<>();
        methodHandlers.put("getAllCellInfo", new GetAllCellInfoMethodHandler());
        methodHandlers.put("getCellLocation", new GetCellLocationMethodHandler());
        methodHandlers.put("listen", new ListenMethodHandler());
        return methodHandlers;
    }

    @Override
    public void replaceBinderProxy(Context context, IBinder proxy) {

    }

    static class GetAllCellInfoMethodHandler extends MethodHandler {

        @Override
        public Object onInvoke(Object originObject, Method method, Object[] args) throws InvocationTargetException, IllegalAccessException {
            if (!GpsMockManager.getInstance().isMocking()) {
                return method.invoke(originObject, args);
            }
            return new ArrayList<CellInfo>();
        }
    }

    static class GetCellLocationMethodHandler extends MethodHandler {

        @Override
        public Object onInvoke(Object originObject, Method method, Object[] args) throws InvocationTargetException, IllegalAccessException {
            if (!GpsMockManager.getInstance().isMocking()) {
                return method.invoke(originObject, args);
            }

            return new GsmCellLocation();
        }
    }

    static class ListenMethodHandler extends MethodHandler {
        @Override
        public Object onInvoke(Object originObject, Method method, Object[] args) throws InvocationTargetException, IllegalAccessException {
            if (!GpsMockManager.getInstance().isMocking()) {
                return method.invoke(originObject, args);
            }
            return null;
        }
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/hook/TelephonyHooker.kt
package com.didichuxing.doraemonkit.hook

import android.content.Context
import android.os.IBinder
import android.telephony.CellInfo
import android.telephony.gsm.GsmCellLocation
import com.didichuxing.doraemonkit.kit.gpsmock.GpsMockManager
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*

/**
 * Created by wanglikun on 2019/4/2
 */
class TelephonyHooker : BaseServiceHooker() {

    override fun serviceName(): String {
        return Context.TELEPHONY_SERVICE
    }

    override fun stubName(): String {
        return "com.android.internal.telephony.ITelephony\$Stub"
    }

    override fun methodHandlers(): MutableMap<String, MethodHandler?> {
        val methodHandlers: MutableMap<String, MethodHandler?> = mutableMapOf()
        methodHandlers["getAllCellInfo"] = GetAllCellInfoMethodHandler()
        methodHandlers["getCellLocation"] = GetCellLocationMethodHandler()
        methodHandlers["listen"] = ListenMethodHandler()
        return methodHandlers
    }

    override fun replaceBinder(context: Context?, proxy: IBinder?) {}
    internal class GetAllCellInfoMethodHandler : MethodHandler {
        @Throws(InvocationTargetException::class, IllegalAccessException::class)
        override fun onInvoke(originObject: Any?, proxyObject: Any?, method: Method?, args: Array<Any>?): Any? {
            return if (!GpsMockManager.instance.isMocking) {
                method!!.invoke(originObject, args)
            } else mutableListOf<CellInfo>()
        }
    }

    internal class GetCellLocationMethodHandler : MethodHandler {
        @Throws(InvocationTargetException::class, IllegalAccessException::class)
        override fun onInvoke(originObject: Any?, proxyObject: Any?, method: Method?, args: Array<Any>?): Any? {
            return if (!GpsMockManager.instance.isMocking) {
                method!!.invoke(originObject, args)
            } else GsmCellLocation()
        }
    }

    internal class ListenMethodHandler : MethodHandler {
        @Throws(InvocationTargetException::class, IllegalAccessException::class)
        override fun onInvoke(originObject: Any?, proxyObject: Any?, method: Method?, args: Array<Any>?): Any? {
            return if (!GpsMockManager.instance.isMocking) {
                method!!.invoke(originObject, args)
            } else null
        }
    }
}